File: petclinic-java-gradle-react-tailwind-ts/src/main/java/dev/pollito/petclinic_java_gradle_react_tailwind_ts/security/AdminUserDetailsService.java
package dev.pollito.petclinic_java_gradle_react_tailwind_ts.security;

import dev.pollito.petclinic_java_gradle_react_tailwind_ts.user.User;
import dev.pollito.petclinic_java_gradle_react_tailwind_ts.user.UserRepository;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class AdminUserDetailsService implements UserDetailsService {

    private final UserRepository userRepository;

    public AdminUserDetailsService(final UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public UserDetails loadUserByUsername(final String username) {
        final User user = userRepository.findByUsernameIgnoreCase(username);
        if (user == null) {
            log.warn("user not found: {}", username);
            throw new UsernameNotFoundException("User " + username + " not found");
        }

        return org.springframework.security.core.userdetails.User.withUsername(username)
                .password(user.getPassword())
                .authorities(user.getUsernameUserRoles().stream()
                        .map(ur -> new SimpleGrantedAuthority(ur.getRole().getName()))
                        .toList())
                .build();
    }

}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/security/AdminUserDetailsService.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.security

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.user.UserRepository
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service


@Service
class AdminUserDetailsService(
    private val userRepository: UserRepository
) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        val user = userRepository.findByUsernameIgnoreCase(username)
        if (user == null) {
            log.warn("user not found: {}", username)
            throw UsernameNotFoundException("User ${username} not found")
        }
        val role = UserRoles.ROLE_UNKNOWN
        val authorities = listOf(SimpleGrantedAuthority(role))
        return User.withUsername(username)
                .password(user.password)
                .authorities(authorities)
                .build()
    }


    companion object {

        val log: Logger = LoggerFactory.getLogger(AdminUserDetailsService::class.java)

    }

}
